airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -83,6 +87,31 @@ def _get_dep_statuses(self, ti, session, dep_context: DepContext):
             session=session,
         )
 
+    @staticmethod
+    def _count_upstreams(ti: "TaskInstance", *, session: "Session"):
+        from airflow.models.taskinstance import TaskInstance
+
+        # Optimization: Don't need to hit the database if no upstreams are mapped.
+        upstream_task_ids = ti.task.upstream_task_ids
+        if ti.task.dag and not any(ti.task.dag.get_task(tid).is_mapped for tid in upstream_task_ids):
```

Comment:
```suggestion
        if ti.task.has_dag() and not any(node.is_mapped for node in ti.task.upstream_list):
```

Reply:
This would be slightly slower (an extra list creation) but I guess it doesn’t trump readability in practical scenarios.

Reply:
This doesn’t work well with Mypy because `upstream_list` returns a list of DAGNode, not AbstractOperator.